New Post: "More variable font options for the macOS `system-ui` font in Chromium 83" (#2961)
GoogleChrome/web.dev: src/site/content/en/blog/catalina-system-ui/index.md

---
title: More variable font options for the macOS `system-ui` font in Chromium 83
subhead: Catalina brings a new united variable system font to macOS 
authors:
  - adamargyle
  - drott
description: Chromium 80 had a system-ui font weight regression on macOS. The reason it broke, and the new super powers post-resolution, are worth the wait in Chromium 83.
date: 2020-05-21
updated: 2020-05-21
hero: hero.jpg
thumbnail: thumb.png

> alt: A bright pink and purple gradient with "macOS Catalina system-ui" going from thin to think from left to right, demonstrating some of the new variation settings
> ---

Reviewer: ```suggestion
alt: A bright pink and purple gradient with "macOS Catalina system-ui" going from thin to think from left to right, demonstrating some of the new variation settings
tags:
  - blog
  - css
  - fonts
---
```

---
title: More variable font options for the macOS `system-ui` font in Chromium 83

> subhead: Catalina brings a new united variable system font to macOS 

Reviewer: ```suggestion
subhead: Catalina brings a new united variable system font to macOS.
```

  </video>
</figure>

### Combining the options
With a few lines of CSS, we can tweak the font settings into a bold of our choice or try out other interesting combinations:

```css
font-weight: 700;
font-weight: bold;
font-variation-settings: 'wght' 750, 'YAXS' 600, 'GRAD' 500, 'opsz' 20;
```

<figure class="w-figure w-figure--fullbleed">
  <video controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/macos-system-ui/system-ui_combined_v2.webm" type="video/webm">
    <source src="https://storage.googleapis.com/web-dev-assets/macos-system-ui/system-ui_combined_v2.mp4">
  </video>
</figure>

And just like that, Chromium users on macOS see your upgraded, custom 750 weight with some fun other tweaks 👍 

## Playground

Click **Remix to Edit** in the Glitch below to get an editable copy of the Glitch, and then edit the new `font-variation-settings` options to see how it affects your font. Remember that this Glitch will only work if you're using a macOS Catalina device.

<div class="glitch-embed-wrap" style="height: 480px; width: 100%;">
  <iframe

>     src="https://glitch.com/embed/#!/embed/macos-system-ui?path=style.css&attributionHidden=true&previewSize=50"

Reviewer: ```suggestion
    src="https://glitch.com/embed/#!/embed/macos-system-ui?path=style.css&attributionHidden=true&previewSize=25"
```
Looks funky, maybe this is better

>     src="https://glitch.com/embed/#!/embed/macos-system-ui?path=style.css&attributionHidden=true&previewSize=25"

Reviewer: ```suggestion
    src="https://glitch.com/embed/#!/embed/macos-system-ui?path=style.css&attributionHidden=true&previewSize=25&sidebarCollapsed=true"
```
Collapses the file tree

    alt="macOS Catalina system-ui sandbox on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

macOS 10.15 added new features to its system font, and in macOS 10.15 a tricky `system-ui` bug was logged in the Chromium bug tracker. I wonder if they are related!?

> {% Aside %}The image for this article was made with CSS and system-ui. Here's the [codepen](https://codepen.io/argyleink/pen/eYpbwWd).{% endAside %}

Reviewer: ```suggestion
{% Aside %}The image for this article was made with CSS and `system-ui`. Here's the [codepen](https://codepen.io/argyleink/pen/eYpbwWd).{% endAside %}
```

> ## Appendix: The 'system-ui' regression {: #regression }

Reviewer: ```suggestion
## Appendix: The `system-ui` regression {: #regression }
```

---
title: More variable font options for the macOS `system-ui` font in Chromium 83
subhead: Catalina brings a new united variable system font to macOS 
authors:
  - adamargyle
  - drott
description: Chromium 80 had a system-ui font weight regression on macOS. The reason it broke, and the new super powers post-resolution, are worth the wait in Chromium 83.
date: 2020-05-21
updated: 2020-05-21
hero: hero.jpg
thumbnail: thumb.png
alt: A bright pink and purple gradient with "macOS Catalina system-ui" going from thin to think from left to right, demonstrating some of the new variation settings
tags:
  - blog
  - css
  - fonts
---

> The ['system-ui' section](https://drafts.csswg.org/css-fonts-4/#system-ui-def) of the CSS Fonts Module Level 4 spec defines a `system-ui` font keyword that allows developers to use the built-in, turbo-optimized, localized, mega-high-quality, no-download-needed, default operating system font font right in their sites and apps. 

Reviewer: ```suggestion
The ['system-ui' section](https://drafts.csswg.org/css-fonts-4/#system-ui-def) of the CSS Fonts Module Level 4 spec defines a `system-ui` font keyword that allows developers to use the built-in, turbo-optimized, localized, mega-high-quality, no-download-needed, default operating system font right in their sites and apps. 
```

```css
body {
  font-family: system-ui;
}
```

> This typography choice is akin to saying:
> 
> {% Blockquote 'browser' %}
> use the native system font for the current locale of this user
> {% endBlockquote %}

Reviewer: ```suggestion
This typography choice is akin to saying "use the native system font for the current locale of this user."
```
This is formatting weird so let's just use normal quotes.

![image](https://user-images.githubusercontent.com/4713486/82603358-5f3d2180-9b67-11ea-9764-4d1642c4389d.png)

On macOS, the `system-ui` font is San Francisco, a font that a design team vetted, tested, and… recently upgraded! First we'll cover the [new exciting variable font features in Catalina](#new-powers), then we'll cover a couple of [bugs and how Chromium engineers resolved them](#regression).

This post assumes that you're already familiar with variable fonts. If not, check out [Introduction to variable fonts on the web](/variable-fonts/) and the video below.

{% YouTube 'B42rUMdcB7c' %}

> ### Browser Compatibility

Reviewer: ```suggestion
### Browser compatibility
```

> At the time of writing, `system-ui` has support from Chromium (since 56), Edge (since 79), Safari (since 11), and from Firefox (since 43) but with the `-apple-system` keyword. See Can I use variable fonts? for updates.

Reviewer: ```suggestion
At the time of writing, `system-ui` has support from Chromium (since 56), Edge (since 79), Safari (since 11), and from Firefox (since 43) but with the `-apple-system` keyword. See [Can I use variable fonts?](https://caniuse.com/#feat=variable-fonts) for updates.
```

> ## New Powers {: #new-powers }

Reviewer: ```suggestion
## New powers {: #new-powers }
```

```css
/* 400-1000 */
font-variation-settings: 'YAXS' 500;
```

<figure class="w-figure w-figure--fullbleed">
  <video controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/macos-system-ui/system-ui_yaxs_v2.webm" type="video/webm">
    <source src="https://storage.googleapis.com/web-dev-assets/macos-system-ui/system-ui_yaxs_v2.mp4">
  </video>
</figure>

### Combining the options
With a few lines of CSS, we can tweak the font settings into a bold of our choice or try out other interesting combinations:

```css
font-weight: 700;
font-weight: bold;
font-variation-settings: 'wght' 750, 'YAXS' 600, 'GRAD' 500, 'opsz' 20;
```

<figure class="w-figure w-figure--fullbleed">
  <video controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/macos-system-ui/system-ui_combined_v2.webm" type="video/webm">
    <source src="https://storage.googleapis.com/web-dev-assets/macos-system-ui/system-ui_combined_v2.mp4">
  </video>
</figure>

And just like that, Chromium users on macOS see your upgraded, custom 750 weight with some fun other tweaks 👍 

> ## Sandbox

Reviewer: ```suggestion
## Playground

Click **Remix to Edit** in the Glitch below to get an editable copy of the Glitch, and then edit the new `font-variation-settings` options to see how it affects your font. Remember that this Glitch will only work if you're using a macOS Catalina device.
```
I think you're right that Sandbox is technically correct here but it seems a bit strange to me because of its associations with security sandboxes. https://en.wikipedia.org/wiki/Sandbox_(software_development)

I added the instructions to make the call-to-action very obvious

<div class="glitch-embed-wrap" style="height: 480px; width: 100%;">
  <iframe

>     src="https://glitch.com/embed/#!/embed/macos-system-ui?path=style.css&attributionHidden=true"

Reviewer: ```suggestion
    src="https://glitch.com/embed/#!/embed/macos-system-ui?path=style.css&attributionHidden=true&previewSize=50"
```
This should show the source code and live app at the same time so that readers don't have to switch back and forth

Reviewer: P.S. I love the Glitch here!

    alt="macOS Catalina system-ui sandbox on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

macOS 10.15 added new features to its system font, and in macOS 10.15 a tricky `system-ui` bug was logged in the Chromium bug tracker. I wonder if they are related!?

> {% Banner 'neutral' %}The image for this article was made with CSS and system-ui, here's the [codepen](https://codepen.io/argyleink/pen/eYpbwWd){% endBanner %}

Reviewer: ```suggestion
{% Aside %}The image for this article was made with CSS and system-ui. Here's the [codepen](https://codepen.io/argyleink/pen/eYpbwWd).{% endAside %}
```
Use of a banner seems a bit funny to me here, aside seems more appropriate

## Appendix: The 'system-ui' regression {: #regression }

{% Aside %}
The rest of this blog post explains how `system-ui` broke in Chromium 80 and how the Chromium engineers fixed it. If this isn't interesting or relevant for you, you can stop reading now.
{% endAside %}

> This story starts with a different bug: [#1005969](https://crbug.com/1005969). This was reported against macOS 10.15 because the system-ui font spacing looked narrow and crammed.

Reviewer: ```suggestion
This story starts with a different bug: [#1005969](https://crbug.com/1005969). This was reported against macOS 10.15 because the `system-ui` font spacing looked narrow and crammed.
```

<figure class="w-figure">
  <img src="./tight-chrome-spacing.png" alt="A comparison of two paragraphs from a Facebook group page. On the left is Chrome and the right is Safari, and Chrome is subtle but slightly tighter in spacing">
  <figcaption class="w-figcaption">Chrome on left (tighter kerning), Safari on right (better optical spacing)</figcaption>

> </figure>

Reviewer: Much better example screenshot. And a clever way to avoid the controversial content. 👍

### Background

Did you ever notice on macOS 10.14 how your paragraphs or headers "snapped" to a different looking font when the size went up or down?

<figure class="w-figure w-figure--fullbleed">
  <video controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/macos-system-ui/system-ui_20to19.webm" type="video/webm">
    <source src="https://storage.googleapis.com/web-dev-assets/macos-system-ui/system-ui_20to19.mp4">
  </video>
</figure>

> On macOS 10.14, the `system-ui` font switched between two fonts depending on the target font size. When text was under `20px`, macOS used "San Francisco Text". When text was `20px` or over, macOS used "San Francisco Display". Optical sizing was built statically into two separate fonts.

Reviewer: ```suggestion
On Mojave (macOS 10.14), the `system-ui` font switched between two fonts depending on the target font size. When text was under `20px`, macOS used "San Francisco Text". When text was `20px` or over, macOS used "San Francisco Display". Optical sizing was built statically into two separate fonts.
```

Catalina (macOS 10.15) shipped a new united variable font for San Francisco. No more managing "Text" and "Display". It also gained the new variation setting [`opsz`](#opsz) described earlier.

```css
h1 {
  font-variation-settings: 'opsz' 20;
}
```

Unfortunately, the default `opsz` value in the new Catalina font is `20`, and Chromium engineers were not prepared to apply `opsz` to the system font. This led to smaller sizes displaying too narrow. 

> To fix that, Chromium needed to apply `opsz` correctly to the system font. This lead to [Issue #1005969](https://crbug.com/1005969) getting fixed. Victory!…?

Reviewer: ```suggestion
To fix that, Chromium needed to apply `opsz` correctly to the system font. This led to [Issue #1005969](https://crbug.com/1005969) getting fixed. Victory! Or was it…?
```

### Not done yet

> This is where it got tricky: Chromium applied `opsz` but something did not look right still. System fonts on Mac have an additional font table called [`trak`](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6trak.html), which tweaks horizontal spacing. While working on the fix, we noticed that on macOS, when retrieving horizontal metrics from a `CTFontRef CoreText` object, the `trak` metrics were already getting factored into the metrics results. Chromium's shaping library [HarfBuzz](https://github.com/harfbuzz/harfbuzz) needs metrics where the `trak` values are not yet factored in.

Reviewer: ```suggestion
This is where it got tricky: Chromium applied `opsz` but something did not look right still. System fonts on Mac have an additional font table called [`trak`](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6trak.html), which tweaks horizontal spacing. While working on the fix, Chromium engineers noticed that on macOS, when retrieving horizontal metrics from a `CTFontRef` object, the `trak` metrics were already getting factored into the metrics results. Chromium's shaping library [`HarfBuzz`](https://github.com/harfbuzz/harfbuzz) needs metrics where the `trak` values are not yet factored in.
```

<figure class="w-figure">
  <img src="./weight-loss.png" alt="A master display of system-ui and all of it's font weight and variations in a list. Half of them have no weight differences applied.">
</figure>

> Internally, [Skia](https://skia.org/) uses both the `CGFontRef` class from [CoreGraphics](https://developer.apple.com/documentation/coregraphics) and the `CTFontRef` class from [CoreText](https://developer.apple.com/documentation/coretext). Due to required internal conversions between those objects (used for keeping backwards compatibility and accessing needed APIs on both classes), Skia would lose weight information in certain circumstances and bold fonts would stop working. This was tracked in [Issue #1057654](https://crbug.com/1057654). 

Reviewer: ```suggestion
Internally, [Skia](https://skia.org/) uses both the `CGFontRef` class from [`CoreGraphics`](https://developer.apple.com/documentation/coregraphics) and the `CTFontRef` class from [`CoreText`](https://developer.apple.com/documentation/coretext). Due to required internal conversions between those objects (used for keeping backwards compatibility and accessing needed APIs on both classes), Skia would lose weight information in certain circumstances and bold fonts would stop working. This was tracked in [Issue #1057654](https://crbug.com/1057654). 
```

> Skia still needs to support macOS 10.11 because Chromium still supports it. On 10.11 the San Francisco Text and San Francisco Display fonts weren't even variable fonts, but each was a family of separate fonts for every weight available. At some point their glyph IDs became out of sync with each other. So if Skia did text shaping (converting text into glyphs that can be drawn) with San Francisco Text, it would be gibberish if drawn with San Francisco Display, and vice versa. And even if Skia just asked for a different size macOS might switch to the other. It should be possible to always use one of the fonts and just scale it (using a matrix to scale it up instead of asking for a larger size) but CoreText has an issue where it will not scale sbix (color emoji) glyphs up (only down). It's a bit more complex than that. CoreText actually seems to cap the vertical extent after matrix application, which seems to be related to it not being able to draw emoji at 45 degree angles. In any event, if you want your emoji to be shown big, you need to make a copy of the font to get a big version.

Reviewer: ```suggestion
Skia still needs to support macOS 10.11 because Chromium still supports it. On 10.11 the "San Francisco Text" and "San Francisco Display" fonts weren't even variable fonts. Rather, each was a family of separate fonts for every weight available. At some point their glyph IDs became out of sync with each other. So if Skia did text shaping (converting text into glyphs that can be drawn) with "San Francisco Text", it would be gibberish if drawn with "San Francisco Display", and vice versa. And even if Skia just asked for a different size macOS might switch to the other. It should be possible to always use one of the fonts and just scale it (using a matrix to scale it up instead of asking for a larger size) but `CoreText` has an issue where it will not scale sbix (color emoji) glyphs up (only down). It's a bit more complex than that. `CoreText` actually seems to cap the vertical extent after matrix application, which seems to be related to it not being able to draw emoji at 45 degree angles. In any event, if you want your emoji to be shown big, you need to make a copy of the font to get a big version.
```

> So in order to create copies of `CTFont` objects at different sizes internally while ensuring that the same underlying font data is used, Chromium pulled the `CGFont` off the `CTFont`, then made a new `CTFont` from the `CGFont` (`CGFont` objects are size independent, the magic switching happens at the CoreText level). This worked fine until 10.154. In 10.15 this round trip ended up losing too much information, resulting in the weight issue. [Flutter noticed the weight issue](https://github.com/flutter/flutter/issues/49492) and an alternate fix for resizing was made to create the new `CTFont` directly from the original `CTFont` while controlling the optical size directly using an old but undocumented attribute in `CoreText`. This keeps things working on 10.11 and fixes other issues (like explicitly setting the optical size to the default value).

Reviewer: ```suggestion
So in order to create copies of `CTFont` objects at different sizes internally while ensuring that the same underlying font data is used, Chromium pulled the `CGFont` off the `CTFont`, then made a new `CTFont` from the `CGFont` (`CGFont` objects are size independent, the magic switching happens at the `CoreText` level). This worked fine until 10.154. In 10.15 this round trip ended up losing too much information, resulting in the weight issue. [Flutter noticed the weight issue](https://github.com/flutter/flutter/issues/49492) and an alternate fix for resizing was made to create the new `CTFont` directly from the original `CTFont` while controlling the optical size directly using an old but undocumented attribute in `CoreText`. This keeps things working on 10.11 and fixes other issues (like explicitly setting the optical size to the default value).
```

> However, this preserves more of the CoreText 'magic' in the font. One of these seems to be that it still tweaks the glyph advances in some way other than just the `trak` table (the application of which Chromium was already trying to suppress through yet another undocumented attribute).

Reviewer: ```suggestion
However, this preserves more of the `CoreText` 'magic' in the font. One of these seems to be that it still tweaks the glyph advances in some way other than just the `trak` table (the application of which Chromium was already trying to suppress through yet another undocumented attribute).
```

> `CGFont` doesn't do any of this 'magic' so maybe Chromium could get the `CGFont` off the `CTFont` and just use it to get advances? Unfortunately this wouldn't work because CoreText is known to muck with fonts in other ways as well. For example, it makes small emoji slightly bigger than you actually requested (boosting their size a bit). `CGFont` doesn't know about this, so you'd end up with your sbix-based emoji too close to each other since you'd be measuring at one size but `CoreText` would draw them bigger by some amount. Chromium does want the `CTFont` advances, but it wants them without tracking, and preferably without any other mucking about.

Reviewer: ```suggestion
`CGFont` doesn't do any of this 'magic' so maybe Chromium could get the `CGFont` off the `CTFont` and just use it to get advances? Unfortunately this wouldn't work because `CoreText` is known to muck with fonts in other ways as well. For example, it makes small emoji slightly bigger than you actually requested (boosting their size a bit). `CGFont` doesn't know about this, so you'd end up with your sbix-based emoji too close to each other since you'd be measuring at one size but `CoreText` would draw them bigger by some amount. Chromium does want the `CTFont` advances, but it wants them without tracking, and preferably without any other mucking about.
```

Since the fix for the spacing issue required a set of interconnected Blink and Skia fixes, the Chromium engineers could not "just revert" to fix the problem. The Chromium engineers also tried using a different build flag for changing a font-related codepath in Skia, which fixed the bold fonts problem, but regressed the spacing problem.

> ## The Fix

Reviewer: ```suggestion
## The fix

```

In the end, of course Chromium wanted to fix both things. Chromium now resorts to using a different implementation for retrieving horizontal metrics directly from the binary data in the system font's font tables. Using this, Chromium is sidestepping `CoreText` and Skia when the font has a `trak` table (except when it's the emoji font). 

<figure class="w-figure">
  <img src="./weight-back.png" alt="A master display of system-ui and all of it's font weight and variations in a list. The half previously not working looks great now.">
</figure>

> In the meantime there's still [Skia Issue #10123](https://bugs.chromium.org/p/skia/issues/detail?id=10123) to track fixing this fully in Skia, and to go back to using Skia to retrieve the system font metrics from there, instead of the current fix that goes through HarfBuzz.

Reviewer: ```suggestion
In the meantime there's still [Skia Issue #10123](https://bugs.chromium.org/p/skia/issues/detail?id=10123) to track fixing this fully in Skia, and to go back to using Skia to retrieve the system font metrics from there, instead of the current fix that goes through `HarfBuzz`.
```
